src/ir/auth_logic/declaration_environment.cc

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright 2022 Google LLC.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "src/ir/auth_logic/declaration_environment.h"
+
+namespace raksha::ir::auth_logic {
+
+namespace {
+// Generates a new mapping from relation names to relation declarations
+// implemented as a flat_hash_map. It is used in the implementation of
+// the constructor for DeclarationEnvironment. This is implemented
+// as a nested class so that it can access the DeclarationEnvironment's
+// private state during the constructor.
+class RelationDeclarationEnvironmentVisitor
+    : public AuthLogicAstTraversingVisitor<
+          RelationDeclarationEnvironmentVisitor> {
+ public:
+  RelationDeclarationEnvironmentVisitor(DeclarationEnvironment& enclosing_env)
+      : enclosing_env_(enclosing_env) {}
+
+ private:
+  Unit Visit(const datalog::RelationDeclaration& rel_decl) {
+    enclosing_env_.AddDeclaration(rel_decl);
+    return Unit();
+  }
+  DeclarationEnvironment& enclosing_env_;
```

Comment:
Why not construct `absl::flat_hash_map<std::string, datalog::RelationDeclaration> inner_map_;` in the visitor and return it?

Reply:
That seems also fine, but then we need to either make the `inner_map_` of the `DeclarationEnvironment` public or give it a public setter method. This seems fine but I don't necessarily see why that is any better or worse than the way it works now.

Reply:
On second thought, fixed this.